File: examples/javascript/graph_1.js
class Graph {
  constructor() { this.adj = new Map(); }
  addVertex(v) { if (!this.adj.has(v)) this.adj.set(v, []); }
  addEdge(u, v) {
    this.addVertex(u); this.addVertex(v);
    this.adj.get(u).push(v);
  }
  dfs(start, visited = new Set()) {
    visited.add(start);
    console.log(start);
    for (const n of this.adj.get(start) || []) {
      if (!visited.has(n)) this.dfs(n, visited);
    }
  }
}

const g = new Graph();
g.addEdge(0, 1); g.addEdge(0, 2); g.addEdge(1, 2); g.addEdge(2, 0); g.addEdge(2, 3);
g.dfs(2);

File: examples/go/graph_1.go
package main

import "fmt"

func dfs(graph map[int][]int, start int, visited map[int]bool) {
	if visited[start] {
		return
	}
	visited[start] = true
	fmt.Println(start)
	for _, n := range graph[start] {
		dfs(graph, n, visited)
	}
}

func main() {
	graph := map[int][]int{
		0: {1, 2},
		1: {2},
		2: {0, 3},
		3: {3},
	}
	visited := make(map[int]bool)
	dfs(graph, 2, visited)
}
